Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -0,0 +1,233 @@
+        layout.invalidateCaches()
+        layout.setupForOrientation(toInterfaceOrientation)
+        collection.setNeedsLayout()
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        datasource = TopSitesDataSource(data: Cursor(status: .Failure, msg: "Nothing loaded yet"))
+
+        layout.registerClass(TopSitesSeperator.self, forDecorationViewOfKind: SeperatorKind)
+
+        collection = UICollectionView(frame: self.view.frame, collectionViewLayout: layout)
+        collection.backgroundColor = UIColor.whiteColor()
+        collection.dataSource = datasource
+        collection.delegate = self
+        collection.registerClass(ThumbnailCell.self, forCellWithReuseIdentifier: ThumbnailIdentifier)
+        collection.registerClass(TopSitesRow.self, forCellWithReuseIdentifier: RowIdentifier)
+        collection.keyboardDismissMode = .OnDrag
+        view.addSubview(collection)
+        collection.snp_makeConstraints { make in
+            make.edges.equalTo(self.view)
+            return
+        }
+    }
+
+    func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
+        if let site = datasource.data[indexPath.item] as? Site {
+            homePanelDelegate?.homePanel(self, didSelectURL: NSURL(string: site.url)!)
+        }
+    }
+}
+
+class TopSitesLayout: UICollectionViewLayout {
+    var numRows: CGFloat = 3
+    var numCols: CGFloat = 2
+    let aspectRatio: CGFloat = 0.7
+
+    var thumbWidth: CGFloat { return CGFloat(width / numCols) }
+    var thumbHeight: CGFloat { return thumbWidth * aspectRatio }
+    var count: Int {
+        if let dataSource = self.collectionView?.dataSource as? TopSitesDataSource {
+            return dataSource.data.count
+        }
+        return 0
+    }
+    var width: CGFloat { return self.collectionView?.frame.width ?? 0 }
+
+    let ToolbarHeight: CGFloat = 44
+    let StatusBarHeight: CGFloat = 20
+    let RowHeight: CGFloat = 70
+
+    var attrCache = [Int: UICollectionViewLayoutAttributes]()
+    var thumbSecHeight: CGFloat { return width / numCols * numRows }
+
+    override init() {
+        super.init()
+        setupForOrientation(UIApplication.sharedApplication().statusBarOrientation)
+    }
+
+    required init(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func setupForOrientation(orientation: UIInterfaceOrientation) {
+        if orientation.isLandscape {
+            numRows = 2
+            numCols = 3
+        } else {
+            numRows = 3
+            numCols = 2
+        }
+    }
+
+    private func getIndexAtPosition(y: CGFloat) -> Int {
+        if y < thumbSecHeight {
+            let row = Int(y / thumbHeight)
+            return min(count-1, max(0, row * Int(numCols)))
+        }
+        return min(count-1, max(0, Int((y - thumbSecHeight) / RowHeight + numRows * numCols)))
```

Comment:
What's going on here? What "index" are we getting, and at what "position"? Comments would help a lot.